contractsV5/eth2dai/Eth2DaiReserve.sol

```diff
@@ -0,0 +1,871 @@
+        maxTakes = minOf(maxTakes, basicData.maxTakes);
+
+        uint minETHAmount = (factorData.minOrderSizeX * ethOrderSize + factorData.minOrderSizeY * PRECISION) / BASIC_FACTOR_STEP;
+
+        // translate min amount to pay token
+        minPayAmount = isEthToToken ? minETHAmount : minETHAmount * order0Pay / order0Buy;
+    }
+
+    function getFirstOffer(ERC20 offerSellGem, ERC20 offerBuyGem)
+        internal view
+        returns(uint offerId, uint offerPayAmount, uint offerBuyAmount)
+    {
+        offerId = otc.getBestOffer(offerSellGem, offerBuyGem);
+        (offerBuyAmount, ,offerPayAmount, ) = otc.getOffer(offerId);
+    }
+
+    function getNextBestOffer(
+        ERC20 offerSellGem,
+        ERC20 offerBuyGem,
+        uint payAmount,
+        uint prevOfferId)
+        public view
+        returns(
+            uint offerId,
+            uint offerPayAmount,
+            uint offerBuyAmount
+        )
+    {
+        if (prevOfferId == INVALID_ID) {
+            offerId = otc.getBestOffer(offerSellGem, offerBuyGem);
+        } else {
+            offerId = otc.getWorseOffer(prevOfferId);
+        }
+
+        (offerBuyAmount, ,offerPayAmount, ) = otc.getOffer(offerId);
+
+        while (payAmount > offerPayAmount) {
+            offerId = otc.getWorseOffer(offerId); // next best offer
+            if (offerId == 0) {
+                offerId = 0;
+                offerPayAmount = 0;
+                offerBuyAmount = 0;
+                break;
+            }
+            (offerBuyAmount, ,offerPayAmount, ) = otc.getOffer(offerId);
+        }
+    }
+
+    function getFirstBidAndAskOrdersPub(ERC20 token)
+        public view
+        returns(uint bidPayAmt, uint bidBuyAmt, uint askPayAmt, uint askBuyAmt)
+    {
+        OfferData memory bid;
+        OfferData memory ask;
+        (bid, ask) = getFirstBidAndAskOrders(token);
+        bidPayAmt = bid.payAmount;
+        bidBuyAmt = bid.buyAmount;
+        askPayAmt = ask.payAmount;
+        askBuyAmt = ask.buyAmount;
+    }
+
+    // bid: buy WETH, ask: sell WETH (their base token is DAI)
+    function getFirstBidAndAskOrders(ERC20 token) internal view returns(OfferData memory bid, OfferData memory ask) {
+        // getting first bid offer (buy WETH)
+        (bid.id, bid.payAmount, bid.buyAmount) = getFirstOffer(token, wethToken);
+        // getting first ask offer (sell WETH)
+        (ask.id, ask.payAmount, ask.buyAmount) = getFirstOffer(wethToken, token);
+    }
+    
+    function checkValidSpread(OfferData memory bid, OfferData memory ask, bool isCheckingMinSpread, uint minSpreadBps)
+        internal pure
+        returns(bool)
+    {
+        // if no bid or ask order, consider as invalid spread?
+        if (bid.id == 0 || ask.id == 0 || bid.buyAmount > MAX_QTY || bid.payAmount > MAX_QTY || ask.buyAmount > MAX_QTY || ask.payAmount > MAX_QTY) {
+            return false;
+        }
+
```

Comment:
all these comments are very good to have.
but not sure it should be public on chain..
it exposes our logics

Reply:
What do you think? I put here just for easier to understand the logics